### PR TITLE
DataFrame [WIP]

### PR DESCRIPTION
**Pull request by gicmo**
This is a test balloon to see where if (I am sure it will) breaks on other platforms (win?).
It is mostly done. Mostly. Need more tests, as always.
Feedback on the API would also be good. Some bits are missing and some are to be discussed:
- Appending rows (currently done via `rows(NEW_SIZE)`)?
- FS backend and resizing of rows.

### CODE REVIEW

File: backend/hdf5/DataFrameHDF5.hpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    DataFrameHDF5(const std::shared_ptr<base::IFile> &file, const std::shared_ptr<base::IBlock> &block, const H5Group &group, const std::string &id, const std::string &type, const std::string &name, time_t time);
+
+
+    void createData(const std::vector<Column> &cols, const Compression &compression);
+
+    std::vector<Column> columns() const override;
+
+    std::vector<unsigned> nameToCol(const std::vector<std::string> &names) const override;
+    std::vector<std::string> colToName(const std::vector<unsigned> &cols) const override;
+
```
Comment (jgrewe) [inline backend/hdf5/DataFrameHDF5.hpp]: I was initially confused by the names, at first sight thought ``nameToCol`` would actually create the column. 

what about names like ``colIndex(vector<string> colNames)`` and ``colNames(vector<unsigned> indices)``

File: include/nix/DataFrame.hpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    Variant readCell(ndsize_t row, const std::string &col) const {
+        std::vector<Variant> cells = backend()->readCells(row, {col});
+        if (cells.size() < 1) {
+            return Variant{};
+        }
+
+        return cells[0];
+    }
+
```
Comment (jgrewe) [inline include/nix/DataFrame.hpp]: would it make sense to offer a function to get the actual ``Cell`` (``vector<Cell> getCell(row, col)``)object?

File: backend/hdf5/DataFrameHDF5.hpp (hunk 1 above)
Comment (gicmo) [reply backend/hdf5/DataFrameHDF5.hpp]: Sounds good.

File: include/nix/DataFrame.hpp (hunk 2 above)
Comment (gicmo) [reply include/nix/DataFrame.hpp]: I thought maybe `getCells(std::vector<Cell> &cells, ndsize_t row, std::vector<unsgined> cols, gboolean resize_vector = true)` ... but one that returns a new vector of `Cell`s could also work.

**Comment by gicmo** [comment]
Todo List:
- [ ] Documentation
- [ ] Group's DataFrame access functions
- [ ] More tests
- [ ] API review and changes

File: include/nix/DataFrame.hpp (hunk 2 above)
Comment (jgrewe) [reply include/nix/DataFrame.hpp]: dunno, both versions seem fine to , except that returning of a new vector of Cells does not imply that the content is a reference to the actually stored cells...

File: include/nix/DataFrame.hpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        DataType dtype = hydra.element_data_type();
+        backend()->writeColumn(col, offset, count, dtype, (const char *) hydra.data());
+    }
+
+    template<typename T>
+        void readColumn(int col,
+                        ndsize_t offset,
+                        std::vector<T> &vals,
+                        ndsize_t count = 0) {
```
Comment (gicmo) [inline include/nix/DataFrame.hpp]: API: This should include a `bool` allowing the vector to be resized to the requested `count` and a way to say "the whole column"

File: include/nix/DataFrame.hpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        Hydra<std::vector<T>> hydra(vals);
+
+        if (count == 0)
+            count = vals.size();
+
+        DataType dtype = hydra.element_data_type();
+        backend()->readColumn(col, offset, count, dtype, (void *) hydra.data());
+    }
+
```
Comment (gicmo) [inline include/nix/DataFrame.hpp]: API: I guess we could expose the raw `readColumn`, `writeColumn` functions too? Or we can create a custom data structure that takes a raw pointer, length and datatype as a proxy structure. Might be nix also for DataArrays. Or both.

File: include/nix/DataFrame.hpp (hunk 3 above)
Comment (jgrewe) [reply include/nix/DataFrame.hpp]: what is the policy now? appending to the vector? maybe put the vector as second arguments and the others get defaults, count == 0 could indicate that the whole column should be read

**Comment by jgrewe** [comment]
Do we also want to offer some sugar? Things like to and from csv? Can be added later on, of course.

File: include/nix/DataFrame.hpp (hunk 3 above)
Comment (gicmo) [reply include/nix/DataFrame.hpp]: currently we use the size of the vector, unless you specify count, then we assume the vector is big enough. We should at least make sure the latter is true I guess ;)

**Comment by gicmo** [comment]
> Do we also want to offer some sugar? Things like to and from csv? Can be added later on, of course.

Yeah, I would add that later. But sounds not like a bad idea. The csv import/export could be used by the FS backend as well.

File: backend/hdf5/DataFrameHDF5.hpp (hunk 1 above)
Comment (gicmo) [reply backend/hdf5/DataFrameHDF5.hpp]: Done.

File: include/nix/DataFrame.hpp (hunk 3 above)
Comment (gicmo) [reply include/nix/DataFrame.hpp]: I have added some commits with a new pair of functions. Maybe have a look if you like them.

**Comment by gicmo** [comment]
Do we want a per-column "type" (not DataType, but as in the second parameter of the DataArray/Block/Group)? A "definition"?

**Comment by achilleas-k** [comment]
Good question. I would expect the DataFrame to have a single type and description attribute like any other object, but since the data isn't homogeneous, it might useful to be able to specify a description for each. But it also gets a bit messy, doesn't it? Having attributes for each column in this way kinda breaks the idea of a unified table object.

I would lean towards not having type and description per column unless I hear a very compelling argument or example.

File: include/nix/DataFrame.hpp (hunk 2 above)
Comment (gicmo) [reply include/nix/DataFrame.hpp]: Maybe `Cell` should inherit from `Variant` and then all `getCell(s)` methods could just return `Cells`.

**Comment by gicmo** [comment]
This is getting close. I guess some more tests are needed, but besides that it is basically ready. Review away ;)

**Comment by gicmo** [comment]
> Is there any straightforward way to delete a specific row or column?

Column might be tricky, I am not sure we can alter it, from the docu:

> When a dataset is created, the storage datatype is defined. After the dataset or attri­bute is created, the datatype cannot be changed.

Rows of course should be doable. You think something like `DataFrame::deleteRow(ndsize_t index)`?

**Comment by achilleas-k** [comment]
Yeah, delete row should be supported like you mentioned, just an index, or a range that deletes and shifts the rest "up".

As for column, yeah, I wasn't thinking implementation and deleting a common would be a pain. Probably easier to create a new one and copy the rest over, which is horrible.

**Pull request merged**